implementations/rust/ockam/ockam_api/src/nodes/service/portals.rs

```diff
@@ -312,6 +312,58 @@ impl NodeManagerWorker {
             }
         })
     }
+
+    pub(super) async fn delete_outlet<'a>(
+        &mut self,
+        req: &Request<'_>,
+        dec: &mut Decoder<'_>,
+    ) -> Result<ResponseBuilder<OutletStatus<'a>>> {
+        let mut node_manager = self.node_manager.write().await;
+        let DeleteOutlet { alias, .. } = dec.decode()?;
+
+        let alias = alias.into_owned();
+        info!("Handling request to delete outlet portal");
+        if let Some(outlet_to_delete) = node_manager.registry.outlets.get(&alias) {
+            let tcp_addr = outlet_to_delete.tcp_addr.clone();
+            let worker_addr = outlet_to_delete.worker_addr.clone();
+            let res = node_manager.tcp_transport.stop_outlet(worker_addr).await;
+
+            Ok(match res {
+                Ok(_) => {
+                    // TODO: Use better way to remove outlets?
+                    let removed_outlet = node_manager.registry.outlets.remove(&alias);
+
+                    if let Some(successfully_removed_outlet) = removed_outlet {
+                        Response::ok(req.id()).body(OutletStatus::new(
+                            successfully_removed_outlet.tcp_addr.to_string(),
+                            successfully_removed_outlet.worker_addr.to_string(),
+                            alias,
+                            None,
+                        ))
+                    } else {
+                        info!("Failed to remove outlet from node manager registry");
+                        Response::internal_error(req.id())
+                            .body(OutletStatus::new(tcp_addr, "", alias, None))
+                    }
+                }
+                Err(e) => Response::bad_request(req.id()).body(OutletStatus::new(
+                    tcp_addr,
+                    "",
+                    alias,
+                    Some(e.to_string().into()),
+                )),
+            })
+        } else {
+            let payload_log = format!("There is no corresponding outlet for the {}", alias);
+            info!(payload_log);
+            Ok(Response::bad_request(req.id()).body(OutletStatus::new(
+                "",
+                "",
+                alias,
+                Some(payload_log.into()),
+            )))
+        }
+    }
```

Comment:
There are some problems in this function:
- You can get and remove in a single operation with `remove_entry`, which helps simplify the code a little
- Removing an item from the registry should always succeed. This is also fixed by applying the previous change
- It's OK to provide some context in the `payload` field, but we shouldn't rely on it to report errors, as this is probably the only response struct that contains that field (see my comment below about the error handling in the command).

Here is a commit I did on a separate branch with the mentioned changes: https://github.com/build-trust/ockam/commit/2e00a734b015165fdb0d96c4ad9d09a1e8f8c7c4

Reply:
Thanks! These look much simpler in the end.

Just a small side note - since we ignore the removed `key` in  `remove_entry` with `_` anyway, shouldn't `remove` also fit the bill here? (since it also returns the removed value). 

I edited the function with your changes, but I kept the `remove` for curiosity's sake and it seems to work out OK as well.

Reply:
Yeah, that's right. The remove returns the entry's value, while remove_entry returns both the key and the value. So, using `remove` is the right choice in this case!